sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/providers/BigQueryStorageWriteApiSchemaTransformProvider.java

```diff
@@ -383,13 +383,6 @@ public PCollectionRowTuple expand(PCollectionRowTuple input) {
         Boolean autoSharding = configuration.getAutoSharding();
         int numStreams = configuration.getNumStreams() == null ? 0 : configuration.getNumStreams();
 
-        // TODO(https://github.com/apache/beam/issues/30058): remove once Dataflow supports multiple
-        // DoFn's per fused step.
-        if (numStreams < 1) {
-          throw new IllegalStateException(
-              "numStreams must be set to a positive integer when input data is unbounded.");
-        }
```

Comment:
This check was added because any Python SDK (or Beam Yaml) user who uses the ReadFromBigQuery with Storage API and unbounded data on Dataflow (at least) will not have data written if `numStreams` is not set to a positive integer. Digging through the logs will show there are warnings about crash loopback due to 2 stateful DoFn's not being able to be fused, but this is not a clear indicator to the issue. This check prevents the job from launching in the first place to avoid user frustration.

Reply:
Runner-specific checks should be done by the runner itself (in this case, this should be stopped on the Dataflow side). We'd rather not restrict all runners.

Reply:
This check is too broad and breaking exisiting tests has been working on direct runner and dataflow:  https://github.com/apache/beam/runs/20839188440 https://github.com/apache/beam/runs/20815247023

On the other hand, there was a similar restriction added to Dataflow side also breaks users and have been rolled back: b/322741233

We were yet to fully understand the cause of "crash loopback ". Using GroupIntoBatches with autosharding alone won't cause issue

Reply:
P.S. when DataflowRunner V2 gets fixed, users will be able to set autosharding. If we restrict it on this Beam version, we unnecessarily mark the version as unusable for autosharding even with a fixed Runner V2.

Reply:
Both really good points. I just wanted there to be a way to fail earlier (unfortunately xlang doesn't pick up runner implementation from what I could see to check dataflow specifically, but sounds like that isn't feasible. LGTM